### PR TITLE
feat: Adding delete metametrics data to security and privacy tab 

### LINKED ISSUE #24406: Adding MetaMetrics data deletion to  Security & Privacy
Part of https://github.com/MetaMask/MetaMask-planning/issues/880
 UI changes for the Delete MetaMetrics Data section with a button.

Comment: We want to update the copy compared to what's currently in Mobile, such that data deletion toggle mentions 30 days instead of 60 days.
IMG_1538

### PR DESCRIPTION
**Pull request by NiranjanaBinoy**
**This  PR is dependant on #24503** 
## **Description**

- Added a new functional component as an entry to the Security & Privacy tab with the `Delete MetaMetrics Data` button.
- A new Delete MetaMetrics Data model will open when you click the button.
- Clicking the `Clear` button in the modal will create a data deletion regulation, update the state, and close the modal, deactivating the `Delete MetaMetrics Data` button.
- The Erroring on the `Clear` button click opens a new error modal.

**Scenarios to disable the DeleteMetaMetrics button:**

1. Metametrics ID not created / not available
2. Just performed a deletion independent on participate in metametrics toggle
3. Participate in metric opt-out & no data is recorded after deletion.
4. Status of current delete regulation as INITIALIZED, RUNNING, or FINISHED and (Participate in metric opt-out/no data recorded after deletion)

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24571?quickstart=1)

## **Related issues**

Fixes #24406, #24407, https://github.com/MetaMask/MetaMask-planning/issues/2523
## **Manual testing steps**

Perquisite:
Provide the following details in the `.metamaskrc` file:
```
ANALYTICS_DATA_DELETION_SOURCE_ID="wygFTooEUUtcckty9kaMc"
ANALYTICS_DATA_DELETION_ENDPOINT="https://proxy.dev-api.cx.metamask.io/segment/v1"
```

1. Make a build(`yarn`, `yarn dist`) against the code.
2. Load the extension in any browser.
3. Navigate to the "Security & privacy" in the Settings 
4. Click on the "Delete MetaMetrics data" button which enables when the "Participate in MetaMetrics" is selected.
5. Validate the post request is made in the service worker with the id - `wygFTooEUUtcckty9kaMc`.

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/app/clear-metametrics-data/ClearMetaMetricsData.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+} from '../../component-library';
+import {
+  AlignItems,
+  BlockSize,
+  Display,
+  FlexDirection,
+  JustifyContent,
+  TextVariant,
+} from '../../../helpers/constants/design-system';
+import { createMetaMetricsDataDeletionTask } from '../../../store/actions';
```
Comment (Gudahtt) [inline ui/components/app/clear-metametrics-data/ClearMetaMetricsData.tsx:31]: Perhaps this PR should stay in draft until the controller exists? As-is this is not mergable because this function does not exist.

File: ui/components/app/clear-metametrics-data/clear-metametrics-data.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      if (participateInMetaMetrics) {
+        await setHasMetaMetricsDataRecorded(true);
+        trackEvent({
+          category: MetaMetricsEventCategory.Settings,
+          event: MetaMetricsEventName.MetricsDataDeletionRequest,
+        });
+      } else {
+        await setHasMetaMetricsDataRecorded(false);
+      }
+    } catch (error: unknown) {
```
Comment (danjm) [inline ui/components/app/clear-metametrics-data/clear-metametrics-data.tsx:64]: Should we also track an error event here? e.g. `MetaMetricsEventName.MetricsDataDeletionRequestError`

cc @gauthierpetetin

File: ui/components/app/clear-metametrics-data/clear-metametrics-data.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+  const closeModal = () => {
+    dispatch(hideDeleteMetaMetricsDataModal());
+  };
+
+  const deleteMetaMetricsData = async () => {
+    try {
+      await createMetaMetricsDataDeletionTask();
+      dispatch(markingMetaMetricsDataDeletion());
+      if (participateInMetaMetrics) {
```
Comment (danjm) [inline ui/components/app/clear-metametrics-data/clear-metametrics-data.tsx:55]: As discussed on call, `trackEvent` does not need to go inside this if block, as `participateInMetaMetrics` is checked at the controller level

File: ui/components/app/clear-metametrics-data/clear-metametrics-data.tsx (hunk 2 above)
Comment (gauthierpetetin) [reply ui/components/app/clear-metametrics-data/clear-metametrics-data.tsx:64]: Hi @danjm ,

In Mobile, we've started to [track expected errors as Analytics](https://github.com/MetaMask/metamask-mobile/blob/main/app/util/metrics/TrackError/trackErrorAsAnalytics.ts) in order to reduce the noise in Sentry, which shall be kept for unexpected errors.
To achieve that we've introduced a generic `Error occured` event that we want to use for all our errors.

Do we have an equivalent generic event for errors in Extension?
- If yes we could use it
- If not, it might be a good occasion to introduce such an event in Extension as well, wdyt?

cc @worldlyjohn @Cal-L for visibility

File: test/e2e/tests/metrics/delete-metametrics-data.spec.ts [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+import {
+  defaultGanacheOptions,
+  withFixtures,
+  getEventPayloads,
+  unlockWallet,
+} from '../../helpers';
+import FixtureBuilder from '../../fixture-builder';
+import { Driver } from '../../webdriver/driver';
+
+const rowLocators = {
```
Comment (hjetpoluru) [inline test/e2e/tests/metrics/delete-metametrics-data.spec.ts:13]: nit: could you please change it to selectors.

**Review by hjetpoluru — COMMENTED** [review]
LGTM! Thanks for your patience.

**Review by hjetpoluru — DISMISSED** [review]

File: privacy-snapshot.json [hunk 5]
```diff
@@ -50,13 +55,12 @@
   "token.api.cx.metamask.io",
   "tokens.api.cx.metamask.io",
   "tx-sentinel-ethereum-mainnet.api.cx.metamask.io",
-  "unresponsive-rpc.url",
-  "www.4byte.directory",
-  "lattice.gridplus.io",
   "unresponsive-rpc.test",
   "authentication.api.cx.metamask.io",
```
Comment (legobeat) [inline privacy-snapshot.json:59]:
```suggestion
```
duplicate

File: privacy-snapshot.json [hunk 6]
```diff
@@ -6,6 +6,7 @@
   "api.web3modal.com",
   "app.ens.domains",
   "arbitrum-mainnet.infura.io",
+  "authentication.api.cx.metamask.io",
```
Comment (legobeat) [inline privacy-snapshot.json:9]:
```suggestion
```
Duplicate (line 59)

File: privacy-snapshot.json [hunk 7]
```diff
@@ -24,19 +25,23 @@
...
+  "lattice.gridplus.io",
   "localhost:8000",
   "localhost:8545",
   "mainnet.infura.io",
   "metamask.eth",
   "metamask.github.io",
+  "metametrics.metamask.test",
   "min-api.cryptocompare.com",
   "nft.api.cx.metamask.io",
+  "oidc.api.cx.metamask.io",
```
Comment (legobeat) [inline privacy-snapshot.json:37]:
```suggestion
```
Duplicate

File: app/_locales/en/messages.json [hunk 8]
```diff
@@ -1547,6 +1547,29 @@
   "deleteContact": {
```
Comment (danjm) [inline app/_locales/en/messages.json:1547]: :heavy_check_mark:

File: shared/constants/metametrics.ts [hunk 9]
```diff
@@ -550,6 +550,7 @@ export enum MetaMetricsEventName {
   MarkAllNotificationsRead = 'Notifications Marked All as Read',
```
Comment (danjm) [inline shared/constants/metametrics.ts:550]: :heavy_check_mark:

File: ui/helpers/constants/settings.js [hunk 10]
```diff
@@ -324,6 +324,13 @@ const SETTINGS_CONSTANTS = [
     route: `${SECURITY_ROUTE}#dataCollectionForMarketing`,
```
Comment (danjm) [inline ui/helpers/constants/settings.js:324]: :heavy_check_mark:

File: ui/helpers/utils/settings-search.test.js [hunk 11]
```diff
@@ -170,7 +170,7 @@ describe('Settings Search Utils', () => {
     it('returns "Security & privacy" section count', () => {
```
Comment (danjm) [inline ui/helpers/utils/settings-search.test.js:170]: :heavy_check_mark:

File: ui/components/app/clear-metametrics-data/clear-metametrics-data.tsx (hunk 2 above)
Comment (danjm) [reply ui/components/app/clear-metametrics-data/clear-metametrics-data.tsx:56]: @NiranjanaBinoy Can you add an 'Error Occured' metrics event, and then track that event here?

File: test/e2e/tests/metrics/delete-metametrics-data.spec.ts [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+        // there is a race condition, where we need to wait before clicking clear button otherwise an error is thrown in the background
+        // we cannot wait for a UI conditon, so we a delay to mitigate this until another solution is found
+        await driver.delay(3000);
+        await driver.clickElementAndWaitToDisappear(selectors.clearButton);
+
+        const deleteMetaMetricsDataButton = await driver.findElement(
+          selectors.deleteMetaMetricsDataButton,
+        );
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        await (deleteMetaMetricsDataButton as any).waitForElementState(
```
Comment (danjm) [inline test/e2e/tests/metrics/delete-metametrics-data.spec.ts:125]: Let's find a way to not use `any` here. What is the type error that you saw that made it necessary to use `any`?

Comment (danjm) [reply test/e2e/tests/metrics/delete-metametrics-data.spec.ts:125]: Same for the other cases

File: ui/components/app/clear-metametrics-data/clear-metametrics-data.test.tsx [hunk 13]
```diff
@@ -0,0 +1,19 @@
+import * as React from 'react';
+import configureStore from '../../../store/store';
+import { renderWithProvider } from '../../../../test/lib/render-helpers';
+
+import ClearMetaMetricsData from './clear-metametrics-data';
+
+describe('ClearMetaMetricsData', () => {
```
Comment (danjm) [inline ui/components/app/clear-metametrics-data/clear-metametrics-data.test.tsx:7]: Can we also add a unit test that checks if there is a button and that `createMetaMetricsDataDeletionTask` is called if the button is clicked?

File: ui/components/app/data-deletion-error-modal/data-deletion-error-modal.test.tsx [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import * as React from 'react';
```
Comment (danjm) [inline ui/components/app/data-deletion-error-modal/data-deletion-error-modal.test.tsx:1]: Can we also add a unit test that checks if there is a button and that `hideDataDeletionErrorModal` is called if the button is clicked?

File: ui/components/app/data-deletion-error-modal/data-deletion-error-modal.tsx [hunk 15]
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
```
Comment (danjm) [inline ui/components/app/data-deletion-error-modal/data-deletion-error-modal.tsx:1]: :heavy_check_mark:

File: ui/ducks/app/app.ts [hunk 16]
```diff
@@ -98,6 +98,8 @@ type AppState = {
   customTokenAmount: string;
```
Comment (danjm) [inline ui/ducks/app/app.ts:98]: :heavy_check_mark:

File: ui/pages/settings/security-tab/__snapshots__/security-tab.test.js.snap [hunk 17]
```diff
@@ -1818,6 +1818,59 @@ exports[`Security Tab should match snapshot 1`] = `
           </label>
```
Comment (danjm) [inline ui/pages/settings/security-tab/__snapshots__/security-tab.test.js.snap:1818]: :heavy_check_mark:

File: ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.test.tsx [hunk 18]
```diff
@@ -0,0 +1,207 @@
+import * as React from 'react';
```
Comment (danjm) [inline ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.test.tsx:1]: :heavy_check_mark:

File: ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.tsx [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+} from '../../../../selectors';
+import { openDeleteMetaMetricsDataModal } from '../../../../ducks/app/app';
+import DataDeletionErrorModal from '../../../../components/app/data-deletion-error-modal';
+import { formatDate } from '../../../../helpers/utils/util';
+import { DeleteRegulationStatus } from '../../../../../shared/constants/metametrics';
+
+const DeleteMetaMetricsDataButton = ({
+  wrapperRef,
+}: {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (danjm) [inline ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.tsx:36]: Let's find a way to remove the use of `any`. What is the error you see that caused you use `any`?

File: ui/pages/settings/security-tab/security-tab.component.js [hunk 20]
```diff
@@ -50,8 +50,10 @@ import {
 } from '../../../helpers/utils/settings-search';
```
Comment (danjm) [inline ui/pages/settings/security-tab/security-tab.component.js:50]: :heavy_check_mark:

File: ui/pages/settings/security-tab/security-tab.container.js [hunk 21]
```diff
@@ -20,10 +20,12 @@ import {
   setUseExternalNameSources,
```
Comment (danjm) [inline ui/pages/settings/security-tab/security-tab.container.js:20]: :heavy_check_mark:

File: ui/pages/settings/security-tab/security-tab.test.js [hunk 22]
```diff
@@ -11,6 +11,8 @@ import { renderWithProvider } from '../../../../test/lib/render-helpers';
 import { getIsSecurityAlertsEnabled } from '../../../selectors';
```
Comment (danjm) [inline ui/pages/settings/security-tab/security-tab.test.js:11]: :heavy_check_mark:

File: privacy-snapshot.json [hunk 23]
```diff
@@ -56,8 +58,6 @@
   "unresponsive-rpc.test",
   "authentication.api.cx.metamask.io",
   "oidc.api.cx.metamask.io",
-  "price.api.cx.metamask.io",
-  "token.api.cx.metamask.io",
```
Comment (NiranjanaBinoy) [inline privacy-snapshot.json:60]: removing duplicate entries

File: ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.tsx [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+} from '../../../../selectors';
+import { openDeleteMetaMetricsDataModal } from '../../../../ducks/app/app';
+import DataDeletionErrorModal from '../../../../components/app/data-deletion-error-modal';
+import { formatDate } from '../../../../helpers/utils/util';
+import { DeleteRegulationStatus } from '../../../../../shared/constants/metametrics';
+
+const DeleteMetaMetricsDataButton = ({
+  wrapperRef,
+}: {
+  wrapperRef?: PolymorphicRef<React.ElementType>;
```
Comment (danjm) [inline ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.tsx:36]: We talked about updating the typing here to match the common practice in the codebase when working with refs

**Review by danjm — CHANGES_REQUESTED** [review]

File: ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.tsx (hunk 24 above)
Comment (NiranjanaBinoy) [reply ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.tsx]: @danjm I have updated these changes  [here]([73f4f14](https://github.com/MetaMask/metamask-extension/pull/24571/commits/73f4f140163f42c3ebc0462c8f24207e498a83f5))

**Review by danjm — DISMISSED** [review]

**Review by hjetpoluru — DISMISSED** [review]
I retested the functionality and its working as expected.

File: test/e2e/tests/metrics/delete-metametrics-data.spec.ts [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+      })),
+  ];
+};
+/**
+ * Scenarios:
+ * 1. Deletion while Metrics is Opted in.
+ * 2. Deletion while Metrics is Opted out.
+ * 3. Deletion when user never opted for metrics.
+ */
+describe('Delete MetaMetrics Data @no-mmi', function (this: Suite) {
```
Comment (DDDDDanica) [inline test/e2e/tests/metrics/delete-metametrics-data.spec.ts:90]: Nit: Is it a lot of effort to migrate this to POM e2e tests pattern? We can also leave it a TODO to be targeted in another PR

File: test/e2e/tests/metrics/delete-metametrics-data.spec.ts [hunk 26]
```diff
@@ -0,0 +1,261 @@
...
+          selectors.deleteMetaMetricsDataButton,
+        );
+        await (
+          deleteMetaMetricsDataButton as WebElementWithWaitForElementState
+        ).waitForElementState('disabled');
+        assert.equal(
+          await deleteMetaMetricsDataButton.isEnabled(),
+          false,
+          'Delete MetaMetrics data button is disabled',
+        );
```
Comment (DDDDDanica) [inline test/e2e/tests/metrics/delete-metametrics-data.spec.ts:132]: Question: it seems like both assertions are waiting for button to be disabled, can we keep only one?

File: test/e2e/tests/metrics/delete-metametrics-data.spec.ts (hunk 25 above)
Comment (NiranjanaBinoy) [reply test/e2e/tests/metrics/delete-metametrics-data.spec.ts:90]: Since we aim to get it merged as soon as possible, I would prefer to add a TODO and create a ticket for the same.

**Review by georgewrmarshall — COMMENTED** [review]
Hey @NiranjanaBinoy, great use of the DS component library! One non-blocking maybe future suggestion would be to create Storybook files for the new modal components, e.g. `clear-metametrics-data.stories.tsx` and `data-deletion-error-modal.stories.tsx`. This would:

- Allow easier review of the UI from designers, engineers, and product managers
- Enable working on the components in isolation
- Simplify future updates to the modals

You could also break the modals into separate PRs to help reduce the size of this one.

Let me know if you need any support with this!

File: ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.test.tsx [hunk 27]
```diff
@@ -0,0 +1,207 @@
...
+
+  it('renders correctly', () => {
+    const store = configureStore({});
+    const { getByText, getByTestId, getAllByText } = renderWithProvider(
+      <DeleteMetaMetricsDataButton />,
+      store,
+    );
+    expect(getByTestId('delete-metametrics-data-button')).toBeInTheDocument();
+    expect(getAllByText('Delete MetaMetrics data')).toHaveLength(2);
+    expect(
```
Comment (DDDDDanica) [inline ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.test.tsx:56]: Optional: you can change this to expect(DOM).toMatchInlineSnapshot() and press `u`, so it is less manual to assert the content

**Comment by DDDDDanica** [comment]
@NiranjanaBinoy the feature looks solid, If you can append a short demo of the deletion feature, it would help a lot to refine and manual testing as well 🙏🏻

**Comment by NiranjanaBinoy** [comment]
@georgewrmarshall I have created a new ticket to add the storybook files, as we are aiming to get this PR merged as soon as possible.
https://github.com/MetaMask/metamask-extension/issues/27630

File: ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.test.tsx (hunk 27 above)
Comment (NiranjanaBinoy) [reply ui/pages/settings/security-tab/delete-metametrics-data-button/delete-metametrics-data-button.test.tsx:56]: updater [06c5d65](https://github.com/MetaMask/metamask-extension/pull/24571/commits/06c5d656f6f6ec921f560d8a1e7141ee187ec6b2)

**Review by DDDDDanica — APPROVED** [review]

**Comment by DDDDDanica** [comment]
LGTM !

**Review by danjm — APPROVED** [review]
